lib/elixir/lib/calendar/date_range.ex

```diff
@@ -0,0 +1,83 @@
+defmodule Date.Range do
+  @moduledoc """
+  Defines a range of dates.
+
+  A range of dates represents a discrete number of dates where
+  the first and last values are dates with matching calendars.
+
+  Ranges of dates can be either increasing (`first <= last`) or
+  decreasing (`first > last`). They are also always inclusive.
+
+  A range of dates implements the `Enumerable` protocol, which means
+  functions in the `Enum` module can be used to work with
+  ranges:
+
+      iex> range = Date.range(~D[2001-01-01], ~D[2002-01-01])
+      iex> Enum.count(range)
+      366
+      iex> Enum.member?(range, ~D[2001-02-01])
+      true
+      iex> Enum.reduce(range, 0, fn(_date, acc) -> acc - 1 end)
+      -366
+
+  """
+
+  @opaque t :: %__MODULE__{first: Date.t, last: Date.t}
```

Comment:
would it make sense to make it a normal type, but only expose `first` and `last` fields (if such thing is allowed)? [1] Otherwise there's no nice way to access these fields without parsing `inspect` output. Alternatively maybe there should be `first/1` and `last/1` functions added?

[1] I found an example in Elixir where there's a difference: https://github.com/elixir-lang/elixir/blob/v1.4.4/lib/elixir/lib/regex.ex#L109:L111 but the reason might be for backwards compat.